subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppCachingIntegrationTest.groovy

```diff
@@ -51,29 +52,116 @@ class CppCachingIntegrationTest extends AbstractCppInstalledToolChainIntegration
+        def newLocation = file('new-location')
+        setupProject(originalLocation)
+        setupProject(newLocation)
+
+        when:
+        inDirectory(originalLocation)
+        withBuildCache().run f.compilationTask
+
+        def snapshotsInOriginalLocation = f.snapshotObjects(originalLocation)
+
+        then:
+        executedAndNotSkipped f.compilationTask
+
+        when:
+        executer.beforeExecute {
+            usingProjectDirectory(newLocation)
+        }
+        run f.compilationTask
+
+        then:
+        executedAndNotSkipped(f.compilationTask)
+        f.assertSameSnapshots(snapshotsInOriginalLocation, f.snapshotObjects(newLocation))
+
+        when:
+        run 'clean'
+        withBuildCache().run f.compilationTask, installTask(buildType)
+
+        then:
+        skipped f.compilationTask
+        f.assertSameSnapshots(snapshotsInOriginalLocation, f.snapshotObjects(newLocation))
+        installation(newLocation.file("build/install/main/${buildType.toLowerCase()}")).exec().out == app.expectedOutput
+
+        where:
+        buildType << [debug, release]
     }
 
-    void compileIsCached() {
-        assert skippedTasks.contains(compilationTask)
-        objectFileFor(file('src/main/cpp/main.cpp'), "build/obj/main/debug").assertExists()
+    String getSourceType() {
+        return 'Cpp'
     }
 
-    void compileIsNotCached() {
-        assert !skippedTasks.contains(compilationTask)
+    class NativeCachingFixture {
+        String buildType
+
+        void assertSameSnapshots(Map<String, TestFile.Snapshot> snapshotsInOriginalLocation, Map<String, TestFile.Snapshot> snapshotsInNewLocation) {
+            assert snapshotsInOriginalLocation.keySet() == snapshotsInNewLocation.keySet()
+            if (nonDeterministicCompilation || absolutePathsInFile) {
+                return
+            }
+            snapshotsInOriginalLocation.each { path, originalSnapshot ->
+                def newSnapshot = snapshotsInNewLocation[path]
+                assert originalSnapshot.hash == newSnapshot.hash
+            }
+        }
+
+        boolean isAbsolutePathsInFile() {
+            buildType == debug || clangOnLinux
+        }
+
+        static boolean isClangOnLinux() {
+            toolChain.displayName == "clang" && OperatingSystem.current().isLinux()
+        }
+
+        Map<String, TestFile.Snapshot> snapshotObjects(TestFile projectDir) {
+            def objDir = projectDir.file('build/obj')
+            def objects = objDir.allDescendants()
+            def snapshots = objects.collectEntries { path ->
+                def obj = objDir.file(path)
+                if (!obj.isFile()) {
+                    return null
+                }
+                [(path): obj.snapshot()]
+            }
+            return snapshots
+        }
+
+        String getCompilationTask() {
+            ":compile${buildType}Cpp"
+        }
+
+        void compileIsCached(TestFile projectDir = temporaryFolder.testDirectory) {
```

Comment:
This doesn't seem to be called with a parameter specified.

Reply:
Now it is.